### ISSUE TITLE
Question about neglogp in PPO2 policies

### ISSUE THREAD

**Issue by pisiiki** [openai/baselines#476]
Hello, I want to write a continuous action space policy for PPO2 with a small constraint. My output should be a vector whose components sum 1.0 and are >=0.0 each.

Until now I have been processing the action inside my env with a softmax because I was unsure about doing this in the policy:

```
a0 = tf.nn.softmax(self.pd.sample(), axis=1)
neglogp0 = self.pd.neglogp(a0)
```

Do I need to write a custom pd object? If that is the case I will appreciate any suggestions.

Thanks.

**Comment by araffin** [comment]
Hello,
by using a softmax on the output, you have changed the distribution, so you need to account for that.
Your problem looks similar to what they do in Soft Actor Critic (https://arxiv.org/abs/1801.01290), see "enforcing action bounds" section.
In the paper, they squash the ouput with a tanh function and used a formula you can find here (https://en.wikipedia.org/wiki/Probability_density_function), in the "Dependent variables and change of variables" section.
However, for your problem, softmax is not bijective (not injective in fact), so it does not seem trivial to adapt to your problem (https://stats.stackexchange.com/questions/7164/change-of-variable-with-non-bijective-function).
A suggestion from that last link is to "divide the domain of your non-bijective function into parts where the function is bijective and then apply change of variables.".

**Comment by pisiiki** [comment]
Hello araffin,

thanks for your answer. It has helped me a lot with some intuitions. I think I found a valid solution. It turns out that my action space is a unit simplex, so I guess I can directly draw from a [Dirichlet distribution](https://en.wikipedia.org/wiki/Dirichlet_distribution) and convert back to cartesian coords. I will need to implement this dist. for baselines btw, but I think that won't be a problem at all.

Regards.
